packages/react-hooks/README.md

```diff
@@ -1,11 +1,301 @@
-# `@w2solutions/react-hooks`
 
-> react hooks helper package
+# @w2solutions/react-hooks
 
-## Usage
 
+ This package includes a set of usefull React hooks.
+
+###  Table of contents
+
+-  [Install](#install)
+
+-  [Hooks](#hooks)
+
+	-  [useAsync](#useasynca-rcallback)
+
+	-  [useDebounceEffect](#usedebounceeffectcallback-delay-dependencyarray)
+
+	-  	[useFileSelect](#usefileselectoptions)
+
+	-  [useFormFields](#useformfieldsinitialvalues-options)
+
+	-  [usePrevious](#useprevioussometypevalue)
+
+	-  [useScrollPosition](#usescrollpositionoptions)  
+
+# Install
+### npm
+```npm i @w2solutions/react-hooks```
+### Yarn
+```yarn add @w2solutions/react-hooks```
+# Hooks
+
+## `useAsync<A, R>(callback)`
+`useAsync` hook lets you call asynchronous functions nice and easy. 
+#### Input
+```javascript
+callback: (...args: A) => Promise<R>
+```
+#### Output
+```javascript
+[
+	(...args: A) => void,
+	{
+		data: R | null,
+		error: Error | null,
+		loading: boolean,
+		calledOnce: boolean
+	}
+]
 ```
-const reactHooks = require('react-hooks');
+#### Example
+```javascript
+import { useAsync } from "@w2solutions/react-hooks"
 
-// TODO: DEMONSTRATE API
+const AsyncInvocation = () => {
+	const [counter, setCounter] = useState(0)
+
+	const asyncFunction = (...args: any[]): Promise<any> => {
+		return new Promise((resolve) => {
+			setTimeout(() => {
+				resolve(args)
+			}, 1000)
+		})
+	}
+
+	const callbackFn = useCallback(() => {
+		return asyncFunction(counter)
+	}, [counter])
+
+	const [callAsyncFunction, stateData] = useAsync(callbackFn)
+
+	return (
+		<>
+			<button onClick={() => setCounter(counter+1)}>
+				Counter +1
+			</button>
+			<button disabled={stateData.calledOnce} onClick={() => callAsyncFunction()}>
+				Call asynchronous function
+			</button>
+			<div style={stateData.loading ? {display: "block"} : {display: "none"}}>
+				Loading...
+			</div>
+			<div>Data: {stateData.data}</div>
+			<div>Error: {stateData.error}</div>
+		</>
+	)
+}
+```
+
+> `asyncFunction` resolves after 1 second. The `asyncFunction` is called within `useCallback` hook, which is passed to the `useAsync` hook. The `calledOnce` flag disables the button, so that the asynchronous function can only be invoked once by this button. `Loading` flag makes a loading div visible.
+
+
+## `useDebounceEffect(callback, delay, dependencyArray)`
+`useDebounceEffect` invokes a callback function after a given delay if any item in the dependency array has been changed.
+#### Input
+```javascript
+callback: EffectCallback
+delay: number
+dependencyArray?: DependencyList
 ```
+#### Example
+```javascript
+import { useDebounceEffect } from "@w2solutions/react-hooks"
+
+const DebounceEffect = () => {
+	const [counter, setCounter] = useState(1)
+	const [anotherCounter, setAnotherCounter] = useState(1)
+
+	const square = useCallback(()=>{
```

Comment:
usage of `useCallback` here should not be required.

`useDebounceEffect` should be the same as `useEffect` with the same examples. 
the useCallback is confusing here since it could mean that it is required for this hook.